### PR TITLE
issue_130_stabilization_guide

### PR DESCRIPTION
**Pull request by rajcspsg**
@mark-i-m @pietroalbini This PR addresses [Stabilization guide] of issue https://github.com/rust-lang/rustc-guide/issues/133.

Please review and let me know your feedback.

### CODE REVIEW

File: src/SUMMARY.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
     - [Build and Install distribution artifacts](./build-install-distribution-artifacts.md)
     - [Documenting Compiler](./compiler-documenting.md)
 - [Coding conventions](./conventions.md)
+- [Stabilizing Features](./stabilization_guide.md)
```
Comment (mark-i-m) [inline src/SUMMARY.md]: Perhaps we could make this an appendix chapter?

File: src/stabilization_guide.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
+
+# Request for stabilization
+
+    Once an unstable feature has been well-tested with no outstanding
```
Comment (mark-i-m) [inline src/stabilization_guide.md]: Could you please un-indent this paragraph? It renders as a code block. (also applies for a few other places below).

File: src/stabilization_guide.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    - Documentation PRs
+    - Write a stabilization report
+    - FCP
+    - Stabilization PR
+
+## Documentation PRs
+
+    Prepare PRs to update documentations involing this new feature.
+    You need to submit PRs for  repositories The Reference, The Book
```
Comment (mark-i-m) [inline src/stabilization_guide.md]: It looks like this used to be a bulleted list?

File: src/stabilization_guide.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    The report should contain:
+
+        - A summary, showing examples (e.g. code snippets) what is
+          enabled by this feature.
+        - Links to test cases in our test suite regarding this feature
+          and describe the feature's behavior on encountering edge cases.
+        - Links to the documentations (the PRs we have made in the
+          previous steps).
+        - Any other relevant information(Examples of such reports can
+          be found in rust-lang/rust#44494 and rust-lang/rust#28237).
```
Comment (mark-i-m) [inline src/stabilization_guide.md]: I would add one other bullet:

- The resolutions of any unresolved questions if the stabilization is for an RFC.

File: src/stabilization_guide.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+Once we have decided to stabilize a feature, we need to have a PR that
+actually makes that stabilization happen. These kinds of PRs are a
+great way to get involved in Rust, as they take you on a little tour
+through the source code.
+
+Here is a general guide to how to stabilize a feature -- every feature
+is different, of course, so some features may require steps beyond 
+what this guide talks about.
+
+Note: Before we stabilize any feature, it is rule that it should appear
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
Note: Before we stabilize any feature, it's the rule that it should appear
```

File: src/stabilization_guide.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+in the documentation.
+
+### Updating the feature-gate listing
+
+There is a central listing of feature-gates in
+`src/libsyntax/feature_gate.rs`. Search for the `declare_features!`
```
Comment (mark-i-m) [inline src/stabilization_guide.md]: Could you make this links to https://doc.rust-lang.org/nightly/nightly-rustc/syntax/feature_gate/index.html

File: src/stabilization_guide.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+macro. There should be an entry for the feature you are aiming to
+stabilize, something like (this example is taken from
+[rust-lang/rust#32409]:
+
+    ```
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
    ```rust
```

File: src/stabilization_guide.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    // pub(restricted) visibilities (RFC 1422)
+    (accepted, pub_restricted, "1.31.0", Some(32409)),
+    // ^^^^^^ note that we changed this
+    ```
+
+Note that, the version number is updated to be the version number
+of the stable release where this feature will appear. This can be
+found by consulting https://forge.rust-lang.org/, which will guide
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
found by consulting [the forge](https://forge.rust-lang.org/), which will guide
```

File: src/stabilization_guide.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+you the next stable release number. You want to add 1 to that,
+because the code that lands today will become go into beta on that
+date, and then become stable after that. So, at the time of this
+writing, the next stable release (what is currently beta, iow) was
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
writing, the next stable release (i.e. what is currently beta) was
```

File: src/stabilization_guide.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+1.30.0, hence I wrote 1.31.0 above.
+
+### Removing existing uses of the feature-gate
+
+Next search for the feature string (in this case, pub_restricted)
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
Next search for the feature string (in this case, `pub_restricted`)
```

File: src/stabilization_guide.md [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+in the codebase to find where it appears. Change uses of
+`#![feature(XXX)]` from the stdlib and rustc crates to be
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
`#![feature(XXX)]` from the `libstd` and any rustc crates to be
```

File: src/stabilization_guide.md [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+     "`pub(restricted)` syntax is experimental");
+    ```
+
+This `gate_feature_post!` macro prints an error if the
+`pub_restricted` feature is not enabled. It is not needed
+now that `#[pub_restricted]` is stable.
+
+For more subtle features, you may find code like this:
+
+    ```
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
    ```rust
```

File: src/stabilization_guide.md (hunk 7 above)
Comment (mark-i-m) [reply src/stabilization_guide.md]: Also, please un-indent this (also applies elsewhere)

File: src/stabilization_guide.md [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    if self.tcx.sess.features.borrow().pub_restricted { /* XXX */ }
+    ```
+
+This `pub_restricted` field (obviously named after the feature)
+would ordinarily be false if the feature flag is not present
+and true if it is. So transform the code to assume that the field
+is true. In this case, that would mean removing the `if` and
+leaving just the `/* XXX */`.
+
+    ```
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
    ```rust
```

File: src/stabilization_guide.md [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    if self.tcx.sess.features.borrow().pub_restricted { /* XXX */ }
+    = becomes ==>
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
    ```
becomes
```rust
```

File: src/stabilization_guide.md [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    /* XXX */
+
+    if self.tcx.sess.features.borrow().pub_restricted && something { /* XXX */ }
+    = becomes ==>
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
    ```
becomes
```rust
```

File: src/stabilization_guide.md [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+    if something { /* XXX */ }
+    ```
+
+## Updating documentation
+
+If any documentation for this feature exists, it should be
+in the `Unstable Book`, located at `src/doc/unstable-book`.
```
Comment (mark-i-m) [inline src/stabilization_guide.md]: Could you make these into links to https://doc.rust-lang.org/unstable-book/index.html and https://github.com/rust-lang/rust/tree/master/src/doc/unstable-book, respectively?

File: src/stabilization_guide.md [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+Regardless of its existence, the page for the feature gate
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
If it exists, the page for the feature gate
```

File: src/stabilization_guide.md [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+should be removed.
+
+If there was documentation there, integrating it into the
+existing documentation is needed.
+
+If there wasn't documentation there, it needs to be added.
+
+Places that may need updated documentation:
+
+    [The Reference]: this must be updated, in full detail.
```
Comment (mark-i-m) [inline src/stabilization_guide.md]: Please make this a bulleted list.

**Comment by mark-i-m** [comment]
Can you also file PR on the forge repo to remove the stabilization guide from their?

**Comment by rajcspsg** [comment]
@mark-i-m I get error `cannot find value xyz in this scope` in place I have the below code example format - 

```rust
// some rust code
```

Please see below - 
https://travis-ci.com/rust-lang/rustc-guide/builds/97640261

It looks like rust tries to compiles the section of code. I tried lot of ways not sure how to fix this.
Any idea?

**Comment by mark-i-m** [comment]
Ah, sorry. I always forget about that. Change the it to ` ```rust,ignore `. This will give us syntax highlighting without trying to compile the code snippet.

**Comment by mark-i-m** [comment]
Also, I'm not sure if you saw them, but it looks like Github collapsed a few comments.

**Comment by rajcspsg** [comment]
Thanks @mark-i-m Yeah I'm aware of it. comments got collapsed after I merged your commit suggestions.

**Comment by mark-i-m** [comment]
I meant that Github also hid 8 more comments above:

image

File: src/stabilization_guide.md (hunk 2 above)
Comment (rajcspsg) [reply src/stabilization_guide.md]: done

File: src/stabilization_guide.md (hunk 3 above)
Comment (rajcspsg) [reply src/stabilization_guide.md]: done

File: src/stabilization_guide.md (hunk 4 above)
Comment (rajcspsg) [reply src/stabilization_guide.md]: done

File: src/stabilization_guide.md (hunk 6 above)
Comment (rajcspsg) [reply src/stabilization_guide.md]: done

File: src/stabilization_guide.md (hunk 7 above)
Comment (rajcspsg) [reply src/stabilization_guide.md]: done

File: src/SUMMARY.md (hunk 1 above)
Comment (mark-i-m) [reply src/SUMMARY.md]: I meant that you could create a new one, i.e. make this a new appendix chapter. All of the appendix chapters are in the src/appendix/ directory, so you would just move this file to that directory and rename it if you wanted to make this an appendix.

File: src/stabilization_guide.md [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+is different, of course, so some features may require steps beyond 
+what this guide talks about.
+
+Note: Before we stabilize any feature, it's the rule that it should appear
+in the documentation.
+
+### Updating the feature-gate listing
+
+There is a central listing of feature-gates in
+`[src/libsyntax/feature_gate.rs]`. Search for the `declare_features!`
```
Comment (mark-i-m) [inline src/stabilization_guide.md]:
```suggestion
[`src/libsyntax/feature_gate.rs`]. Search for the `declare_features!`
```

Also, you will probably need to update the definition at the bottom of the target.

**Comment by rajcspsg** [comment]
@mark-i-m is it possible to rebuild this https://travis-ci.com/rust-lang/rustc-guide/builds/97764368?

**Comment by rajcspsg** [comment]
Thanks @mark-i-m I've incorporated your comment. Please review and let me know if any changes are required.

**Comment by mark-i-m** [comment]
Thanks!

**Comment by rajcspsg** [comment]
@mark-i-m do I need to send PR to rust forge to remove stabilization comment?

**Comment by rajcspsg** [comment]
Thanks @mark-i-m :)
